dashboard/plugins/s3_usage/s3_stats.py

```diff
@@ -0,0 +1,193 @@
+import boto3
+import sqlite3
+import fcntl
+import os
+import time
+import threading
+import re
+from collections import defaultdict
+from concurrent.futures import ThreadPoolExecutor
+from dataclasses import dataclass
+from dashboard.utils import sizeof_fmt
+
+class S3Stats:
+    ORDER_DEFINITION = {
+        'alphabetically': 'path asc',
+        'avg_file_size_desc': 'size/files asc',
+        'files_desc': 'files desc',
+        'size_desc': 'size desc'
+    }
+    def __init__(self, logger, params):
+        self.logger = logger
+        self.params = params
+        self.validate_params()
+
+    def validate_params(self):
+        if not 'ttl' in self.params:
+            self.params['ttl'] = 60 * 60 * 24 # 1 day
+
+    def is_ready(self):
+        return os.path.exists(S3StatsRefreshTask.STATS_DB_PATH)
+
+    def get_usage_by_storage_class(self, path):
+        data = self.query(path, children_only=False, depth=len(path.split('/'))-1)
+        if len(data) == 0:
+            return {}
+
+        return {
+            sclass: {
+                'size': sizeof_fmt(data[0][sclass]),
+                'percent': data[0][sclass] / data[0]['size']
+            } for sclass in ['size_standard', 'size_ia', 'size_glacier']
+        }
+        
+
+    def query(self, base, depth=None, order="size_desc", children_only=False):
+        if depth is None:
+            depth = len(base.split('/'))
+        if type(depth) is tuple:
+            depth_condition = f'depth between {depth[0]} and {depth[1]}'
+        else:
+            depth_condition = f'depth = {depth}'
+
+        if children_only:
+            path_condition = f"path like '{base}/%'"
+        else:
+            path_condition = f"(path like '{base}/%' or path = '{base}')"
+
+        sqlite = sqlite3.connect(S3StatsRefreshTask.STATS_DB_PATH)
+        sqlite.row_factory = sqlite3.Row
+        cursor = sqlite.cursor()
+        order = self.ORDER_DEFINITION.get(order, self.ORDER_DEFINITION['size_desc'])
+        query = f"SELECT *, (size_standard + size_ia + size_glacier) as size FROM stats WHERE {path_condition} and {depth_condition} order by {order}"
+        self.logger.debug(f"S3Stats: Executing query: {query}")
+        cursor.execute(query)
+        result = cursor.fetchall() 
+        cursor.close()
+        sqlite.close()
+        return result
+
+    def run_refresher_task(self):
+        try:
+            S3StatsRefreshTask(self.logger, **(self.params)).run()
+            next_run_after = self.params['ttl']
+        except:
```

Comment:
This is a very generic except block. Can we make it more specific? What kind of exception is thrown here?

Reply:
This is intentional - I want to catch any exception that the refresh task may throw to make sure that the background scheduling will be not affected. I will add more logging here to not loose the exception context